docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,162 @@
+	```
+- Store only the child node key for the below version in node body writes. Because it is possible to get the child path for the same version.
+	```go
+	func (node *Node) getLeftNode() (*Node, error) {
+		if node.leftNode != nil {
+			return node.leftNode, nil
+		}
+		if node.leftNodeKey != nil {
+			return getNode(node.leftNodeKey) // get the node from the storage
+		}
+		return getNode(&NodeKey{
+			version: node.nodeKey.version,
+			path: node.nodeKey.path + '0',  // it will be more complicated in the real implementation
+		})
+	}
+	```
+- Remove the `version` field from node body writes.
+- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field in the node body.
+- Remove the `orphans` completely from both tree and storage.
+
+New node structure
+
+```go
+type NodeKey struct {
+    version int64
+    path    []byte
+}
+
+type Node struct {
+	key           []byte
+	value         []byte
+	hash          []byte     // keep it in the storage instead of leftHash and rightHash
+	nodeKey       *NodeKey   // new field, the key in the storage
+	leftNodeKey   *NodeKey   // new field, need to store in the storage
+	rightNodeKey  *NodeKey   // new field, need to store in the storage
+	leftNode      *Node
+	rightNode     *Node
+    	size          int64
+	subtreeHeight int8
+}
+```
+
+New tree structure
+
+```go
+type MutableTree struct {
+	*ImmutableTree                                    
+	lastSaved                *ImmutableTree
+	versions                 map[int64]bool           
+	allRootLoaded            bool                     
+	unsavedFastNodeAdditions map[string]*fastnode.Node
+	unsavedFastNodeRemovals  map[string]interface{}   
+	ndb                      *nodeDB
+	skipFastStorageUpgrade   bool 
+
+	mtx sync.Mutex
+}
+```
+
+We will assign the `nodeKey` when saving the current version in `SaveVersion`. It will reduce unnecessary checks in CRUD operations of the tree and keep sorted the order of insertion in the LSM tree.
+
+### Migration
+
+We can migrate nodes one by one by iterating the version.
+
+- Iterate the version in order, and get the root node for the specific version.
+- Iterate the tree based on the root and pick only nodes the node version is the same as the given version.
+- Store them using the new node key format.
+
+### Pruning
+
+We assume keeping only the range versions of `fromVersion` to `toVersion`. Refer to [this issue](https://github.com/cosmos/cosmos-sdk/issues/12989).
+
+When we want to prune all versions up to the specific version `n`
+
+- Iterate the tree based on the root of `n+1`th version.
+- Iterate the node until visiting the node the version is below `fromVersion` and don't visit further deeply.
+- Apply `DeletePath` for all visited nodes the version is below `n+1`.
+
+```go
+func DeletePath(nk *NodeKey) error {
+	DeleteNode(node)
+	if nk.path is not root {
+		DeletePath(&NodeKey{
+			version: nk.version,
+			path: parent(nk.path), // it looks like removing the last binary
```

Comment:
It can only find the parent in the version the node is created in, but there could be multiple parents to delete in the range of versions to be deleted, this method don't delete all orphaned nodes.

Reply:
the multiple parents would be deleted when the same version child is deleted.

Reply:
With the method you described, it'll only delete one parent of the node, right?

Reply:
why only the parent? the purpose of this function is to delete all parents of the path to the root.

Reply:
@yihuang , you are right. There is a case that has two lower version children when rebalancing the tree, for those nodes, they will are not removed in `DeletePath`. I introduced a new way to get orphaned nodes. This is a little tricky to implement but its idea itself is clear. Please focus on removing `orphans`. It will give us more flexibility.